### PR TITLE
OCPBUGS-1430: Add assigned egress ips into capacity

### PR DESCRIPTION
**Pull request by pperiyasamy**
This makes sure `capacity` field in `cloud.network.openshift.io/egress-ipconfig `node annotation (which is introduced in 4.10) set with correct value when cluster upgrade happens from 4.9 and node is also already assigned with egress ips. Otherwise `cloud.network.openshift.io/egress-ipconfig ` annotation contains capacity set as `instanceCapacity-currentUsage` (which excludes assigned egress ips) and this creates problem in configuring existing egress ips on sdn node with below error.

```
E0718 10:51:22.032527 2146 egressip.go:254] Ignoring invalid HostSubnet ip-10-24-205-88.eu-west-1.compute.internal (host: "ip-10-24-205-88.eu-west-1.compute.internal", ip: "10.24.205.88", subnet: "10.128.12.0/23"): the amount of requested EgressIPs (8) on hostSubnet: "ip-10-24-205-88.eu-west-1.compute.internal" exceeds the node's capacity (6)
E0718 10:51:22.035593 2146 egressip.go:254] Ignoring invalid HostSubnet ip-10-24-205-19.eu-west-1.compute.internal (host: "ip-10-24-205-19.eu-west-1.compute.internal", ip: "10.24.205.19", subnet: "10.129.14.0/23"): the amount of requested EgressIPs (9) on hostSubnet: "ip-10-24-205-19.eu-west-1.compute.internal" exceeds the node's capacity (5)
```

Signed-off-by: Periyasamy Palanisamy <pepalani@redhat.com>

### CODE REVIEW

File: pkg/controller/node/node_controller.go
Comment (andreaskaris) [inline pkg/controller/node/node_controller.go:123]: nit (I know it's pedantic): "error listing cloud private ip config, err: %v"

**Review by andreaskaris — CHANGES_REQUESTED** [review]

File: pkg/controller/node/node_controller.go
Comment (pperiyasamy) [reply pkg/controller/node/node_controller.go:143]: yes @andreaskaris , you are correct. i too just noticed providers other than aws use IP field for capacity, so this must be dealt at provider plugin itself with suggested interface change. Will update the PR with required changes.

**Comment by flavio-fernandes** [comment]
/hold
Per: https://github.com/openshift/cloud-network-config-controller/pull/69#discussion_r998268353

File: pkg/controller/node/node_controller.go
Comment (pperiyasamy) [reply pkg/controller/node/node_controller.go]: done

File: pkg/controller/node/node_controller.go
Comment (pperiyasamy) [reply pkg/controller/node/node_controller.go]: sure, added new method for this.

File: pkg/controller/node/node_controller.go
Comment (pperiyasamy) [reply pkg/controller/node/node_controller.go]: correct, nice thought. moved this method into new package and made it to return ip family as well.

File: pkg/controller/node/node_controller.go
Comment (pperiyasamy) [reply pkg/controller/node/node_controller.go]: Thanks again for bringing this up. made the relevant changes.

File: pkg/cloudprovider/openstack.go
Comment (andreaskaris) [inline pkg/cloudprovider/openstack.go:454]: Can you pass the cloudPrivateIPConfigs in here, instead of the length?

File: pkg/cloudprovider/openstack.go
Comment (andreaskaris) [inline pkg/cloudprovider/openstack.go:490]: It's a bit odd in OSP, but I decided to report the capacity per OSP interface.  So, pass in cloudPrivateIPConfigs, and then in line 490 make sure that the cloudPrivateIP is acutally part of the ipnet:
~~~
        var cloudPrivateIPsCount int

	// Loop over all subnets. OpenStack potentially has several IPv4 or IPv6 subnets per port, but the
	// CloudPrivateIPConfig expects only a single subnet of each address family per port. Throw an error
	// in such a case.
	for _, s := range subnets {
		// Parse CIDR information into ip and ipnet.
		ip, ipnet, err = net.ParseCIDR(s.CIDR)
		if err != nil {
			return nil, fmt.Errorf("could not parse subnet information %s for network %s, err: %q",
				s.CIDR, p.NetworkID, err)
		}
		// For IPv4 and IPv6, calculate the capacity.
		if utilnet.IsIPv4(ip) {
			if ipv4 != "" {
				return nil, fmt.Errorf("found multiple IPv4 subnets attached to port %s, this is not supported", p.ID)
			}
			ipv4 = ipnet.String()
		} else {
			if ipv6 != "" {
				return nil, fmt.Errorf("found multiple IPv6 subnets attached to port %s, this is not supported", p.ID)
			}
			ipv6 = ipnet.String()
		}
                // Loop over all cloudPrivateIPConfigs and check if they are part of this ipnet.
                // If the IP is contained in the ipnet, increase the counter.
                for _, cpic := range cloudPrivateIPConfigs {
                  cip, _ := cloudprivateipconfig.CloudPrivateIPConfigNameToIP(cpic.name)
                  if ipnet.Contains(cip) {
                    cloudPrivateIPsCount++
                 }
                }
	}

        c := openstackMaxCapacity + cloudPrivateIPsCount - len(p.AllowedAddressPairs)
~~~
something like that

File: pkg/cloudprovider/openstack_test.go
Comment (andreaskaris) [inline pkg/cloudprovider/openstack_test.go:760]: Modify the unit test:
~~~
tcs := []struct {
		port               neutronports.Port
		nodeEgressIPConfig NodeEgressIPConfiguration
                cloudPrivateIPConfigs []*v1.CloudPrivateIPConfig
		errString         
(...) 
~~~

Then, pass that in as a value: tc.cloudPrivateIPConfigs.

Modify the second test case (the first one and the second one are redundant O_o my bad); include one cloudPrivateIPConfig that's on 192.0.2.0/2, one cloudPrivateIPConfig that's on 2000::/64, and one cloudPrivateIPConfig that's on neither of those subnets.   Adjust:: `IP: openstackMaxCapacity - 2,`  to be `IP: openstackMaxCapacity,` as the capacity should be offset by the 2 matches, but not by the 3rd IP.

Comment (andreaskaris) [reply pkg/cloudprovider/openstack_test.go:760]: I can also do that in a follow-up PR if you don't want to mess with the OSP stuff.

File: pkg/controller/node/node_controller.go
Comment (andreaskaris) [inline pkg/controller/node/node_controller.go:130]: Oh, that took me a while:
You're just reusing the cloudPrivateIPConfigs slice, overwriting the lower keys, and then passing a slice with the index's length. Perhaps document what you're doing with a comment?  (but perhaps it's just me, that's really optional)

**Review by andreaskaris — CHANGES_REQUESTED** [review]
Minor things and OSP (which I can also address in a follow-up PR if you prefer).
Are you going to squash both commits together?
Also, @kyrtapz

File: pkg/cloudprovider/openstack.go
Comment (pperiyasamy) [reply pkg/cloudprovider/openstack.go:502]: sure, done.

File: pkg/cloudprovider/openstack_test.go
Comment (pperiyasamy) [reply pkg/cloudprovider/openstack_test.go]: Thanks for detailed explanation about OSP, modified the test to cover this as well.

**Review by pperiyasamy — COMMENTED** [review]
@andreaskaris  squashed into single commit now.

File: pkg/cloudprivateipconfig/cloudprivateipconfig.go
Comment (andreaskaris) [inline pkg/cloudprivateipconfig/cloudprivateipconfig.go:8]: nit: cloudprivateipconfig.go:9:6: exported type IPFamily should have comment or be unexported

File: pkg/cloudprivateipconfig/cloudprivateipconfig.go
Comment (andreaskaris) [inline pkg/cloudprivateipconfig/cloudprivateipconfig.go:12]: nit: cloudprivateipconfig.go:12:2: comment on exported const IPv4 should be of the form "IPv4 ..."

File: pkg/cloudprivateipconfig/cloudprivateipconfig.go
Comment (andreaskaris) [inline pkg/cloudprivateipconfig/cloudprivateipconfig.go:13]: nit: cloudprivateipconfig.go:14:2: exported const IPv6 should have comment (or a comment on this block) or be unexported

File: pkg/cloudprivateipconfig/cloudprivateipconfig.go
Comment (andreaskaris) [inline pkg/cloudprivateipconfig/cloudprivateipconfig.go:29]: nit: cloudprivateipconfig.go:29:6: func name will be used as cloudprivateipconfig.CloudPrivateIPConfigNameToIP by other packages, and that stutters; consider calling this NameToIP

(that's what the linter gives me, feel free to ignore)

**Review by andreaskaris — COMMENTED** [review]
A few nits after running my linter

Also, can you actually make the modifications to go.mod, go.sum and the vendors directory an individual commit
I just realized that your modification with vendors has ~180000 lines
so having one commit with your changes and one with the vendored changes ... that's also help backporting if we need it I guess
and thanks for all the changes

**Comment by kyrtapz** [comment]
Egress IP was not supported on cloud prior to 4.10 so there was an assumption that there will be [no egress IPs assigned to the node when cncc starts](https://github.com/openshift/cloud-network-config-controller/pull/69/files#diff-a6af1de71edb43b701d83b9ee10e345cf3c6ef3fc738ee3132c3309b6958896dL57).

The capacity is only set [once](https://github.com/openshift/cloud-network-config-controller/pull/69/files#diff-92f10e93215b8da12723d913e0985568114aab4b8afb67541c2a3a9ec5cafba6L103) which I think may race with setting the status of every IP that was assigned to the VM prior to the upgrade. Were you able to verify that this is not possible?

Overall I think the proposed change is good as it provides more resiliency(e.g random annotation removal) but I do not think that it protects us from all capacity mismatch issues.

**Comment by pperiyasamy** [comment]
> Egress IP was not supported on cloud prior to 4.10 so there was an assumption that there will be [no egress IPs assigned to the node when cncc starts](https://github.com/openshift/cloud-network-config-controller/pull/69/files#diff-a6af1de71edb43b701d83b9ee10e345cf3c6ef3fc738ee3132c3309b6958896dL57).
> 
> The capacity is only set [once](https://github.com/openshift/cloud-network-config-controller/pull/69/files#diff-92f10e93215b8da12723d913e0985568114aab4b8afb67541c2a3a9ec5cafba6L103) which I think may race with setting the status of every IP that was assigned to the VM prior to the upgrade. Were you able to verify that this is not possible?

@kyrtapz In fact egress ip was supported in 4.9 as per documentation [here](https://docs.openshift.com/container-platform/4.9/networking/openshift_sdn/assigning-egress-ips.html), bug [OCPBUGS-1430](https://issues.redhat.com/browse/OCPBUGS-1430) is also about upgrading from 4.9 to 4.10 while egress ips already configured in the node. The `cloud.network.openshift.io/egress-ipconfig` annotation is introduced only in 4.10 and ovnk master adds node for egress ip allocation based on this annotation, so i guess we are safe there.

**Comment by kyrtapz** [comment]
> > Egress IP was not supported on cloud prior to 4.10 so there was an assumption that there will be [no egress IPs assigned to the node when cncc starts](https://github.com/openshift/cloud-network-config-controller/pull/69/files#diff-a6af1de71edb43b701d83b9ee10e345cf3c6ef3fc738ee3132c3309b6958896dL57).
> > The capacity is only set [once](https://github.com/openshift/cloud-network-config-controller/pull/69/files#diff-92f10e93215b8da12723d913e0985568114aab4b8afb67541c2a3a9ec5cafba6L103) which I think may race with setting the status of every IP that was assigned to the VM prior to the upgrade. Were you able to verify that this is not possible?
> 
> @kyrtapz In fact egress ip was supported in 4.9 as per documentation [here](https://docs.openshift.com/container-platform/4.9/networking/openshift_sdn/assigning-egress-ips.html), bug [OCPBUGS-1430](https://issues.redhat.com/browse/OCPBUGS-1430) is also about upgrading from 4.9 to 4.10 while egress ips already configured in the node. The `cloud.network.openshift.io/egress-ipconfig` annotation is introduced only in 4.10 and ovnk master adds node for egress ip allocation based on this annotation, so i guess we are safe there.

https://docs.openshift.com/container-platform/4.9/networking/openshift_sdn/assigning-egress-ips.html
```
Egress IPs on Amazon Web Services (AWS), Google Cloud Platform (GCP), and Azure are supported only on OpenShift Container Platform version 4.10 and later.
```

**Comment by pperiyasamy** [comment]
> Egress IPs on Amazon Web Services (AWS), Google Cloud Platform (GCP), and Azure are supported only on OpenShift Container Platform version 4.10 and later.

Ah! I see. Thanks. will make a note of it in the bug.

File: pkg/cloudprovider/openstack_test.go
Comment (andreaskaris) [inline pkg/cloudprovider/openstack_test.go:748]: Ironically, this shows that we can now get above the default / max capacity in corner cases ;-) On the other hand, I suppose this shouldn't happen in real life ... otherwise, we'd have to have a math.Max() for all the cloudprovider plugins where the absolute highest value that we report is the port's default capacity?

Comment (pperiyasamy) [reply pkg/cloudprovider/openstack_test.go:748]: yes, anyway default capacity is always `openstackMaxCapacity - len(p.AllowedAddressPairs)`, should we just ignore counting`cloudPrivateIPsCount` for OSP case ?

**Comment by andreaskaris** [comment]
Overall, this looks good to me now. I haven't had a thorough second look, though and might look at this again, later. I'll try this on a couple of clouds to be sure

**Comment by andreaskaris** [comment]
Checked on GCP:
~~~
[akaris@linux 2104784]$ oc get -o yaml node ci-ln-4rlmcrt-72292-5q4cb-worker-a-727v9 | grep egress
oc     cloud.network.openshift.io/egress-ipconfig: '[{"interface":"nic0","ifaddr":{"ipv4":"10.0.128.0/17"},"capacity":{"ip":10}}]'
e[akaris@linux 2104784]$ oc edit nodeci-ln-4rlmcrt-72292-5q4cb-worker-a-727v99
node/ci-ln-4rlmcrt-72292-5q4cb-worker-a-727v9 edited
[akaris@linux 2104784]$ oc get -o yaml node ci-ln-4rlmcrt-72292-5q4cb-worker-a-727v9 | grep egress
k8s.ovn.org/egress-assignable: ""
[akaris@linux 2104784]$ oc delete pod -n openshift-cloud-network-config-controller          cloud-network-config-controller-57645db5cc-2sqhk
pod "cloud-network-config-controller-57645db5cc-2sqhk" deleted
[akaris@linux 2104784]$ oc get pods -A -o wide | grep cloud-network-config
openshift-cloud-network-config-controller          cloud-network-config-controller-57645db5cc-dv582                    1/1     Running     0              5s     10.128.0.36   ci-ln-4rlmcrt-72292-5q4cb-master-0         <none>           <none>
[akaris@linux 2104784]$ oc logs -f -n openshift-cloud-network-config-controller          cloud-network-config-controller-57645db5cc-dv582
(...)
I1026 17:59:27.315056       1 node_controller.go:146] Setting annotation: 'cloud.network.openshift.io/egress-ipconfig: [{"interface":"nic0","ifaddr":{"ipv4":"10.0.128.0/17"},"capacity":{"ip":10}}]' on node: ci-ln-4rlmcrt-72292-5q4cb-worker-a-727v9
I1026 17:59:27.347691       1 controller.go:160] Dropping key 'ci-ln-4rlmcrt-72292-5q4cb-worker-a-727v9' from the node workqueue
^C
[akaris@linux 2104784]$ oc get cloudprivateipconfig
NAME          AGE
10.0.129.11   2m33s
10.0.129.12   111s
[akaris@linux 2104784]$ oc get egressip
NAME         EGRESSIPS     ASSIGNED NODE                              ASSIGNED EGRESSIPS
egressip11   10.0.129.11   ci-ln-4rlmcrt-72292-5q4cb-worker-a-727v9   10.0.129.11
egressip12   10.0.129.12   ci-ln-4rlmcrt-72292-5q4cb-worker-a-727v9   10.0.129.12
~~~

**Comment by pperiyasamy** [comment]
> Checked on GCP:

Have also tested it in AWS and Azure platforms and fix is working as expected.

**Comment by kyrtapz** [comment]
/approve
@andreaskaris please lgtm if it looks good to you, especially for OSP

**Comment by andreaskaris** [comment]
OpenStack looks good, too:
~~~
[akaris@linux egressip]$ oc get egressip
NAME       EGRESSIPS     ASSIGNED NODE                        ASSIGNED EGRESSIPS
egressip   10.0.129.10   6gks8w2-b5564-rw5lb-worker-0-flh8l   10.0.129.10
[akaris@linux egressip]$ oc get nodes 6gks8w2-b5564-rw5lb-worker-0-flh8l -o yaml | grep egress
    cloud.network.openshift.io/egress-ipconfig: '[{"interface":"508fc058-3044-459f-9e00-a42aab19b57b","ifaddr":{"ipv4":"10.0.0.0/16"},"capacity":{"ip":8}}]'
    k8s.ovn.org/egress-assignable: ""
[akaris@linux egressip]$ oc edit node 6gks8w2-b5564-rw5lb-worker-0-flh8l
node/6gks8w2-b5564-rw5lb-worker-0-flh8l edited
[akaris@linux egressip]$ oc get nodes 6gks8w2-b5564-rw5lb-worker-0-flh8l -o yaml | grep egress
    k8s.ovn.org/egress-assignable: ""
[akaris@linux egressip]$ oc get pods -A -o wide | grep cloud-network-config-contro
openshift-cloud-network-config-controller          cloud-network-config-controller-6fd6f4f589-sb6jh              1/1     Running     1 (22m ago)     33m     10.129.0.32   6gks8w2-b5564-rw5lb-master-0         <none>           <none>
[akaris@linux egressip]$ oc delete pod -n openshift-cloud-network-config-controller          cloud-network-config-controller-6fd6f4f589-sb6jh
pod "cloud-network-config-controller-6fd6f4f589-sb6jh" deleted
[akaris@linux egressip]$ sleep 120 ; oc get nodes 6gks8w2-b5564-rw5lb-worker-0-flh8l -o yaml | grep egress
    k8s.ovn.org/egress-assignable: ""
[akaris@linux egressip]$ oc get pods -n openshift-cloud-network-config-controller
NAME                                               READY   STATUS    RESTARTS   AGE
cloud-network-config-controller-6fd6f4f589-dqhqr   1/1     Running   0          5m35s
[akaris@linux egressip]$ oc logs -n openshift-cloud-network-config-controller cloud-network-config-controller-6fd6f4f589-dqhqr
W1104 11:24:41.516296       1 client_config.go:617] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I1104 11:24:41.517562       1 leaderelection.go:248] attempting to acquire leader lease openshift-cloud-network-config-controller/cloud-network-config-controller-lock...
I1104 11:27:08.581552       1 leaderelection.go:258] successfully acquired lease openshift-cloud-network-config-controller/cloud-network-config-controller-lock
I1104 11:27:08.582270       1 openstack.go:136] Could not find custom CA bundle in file '/kube-cloud-config/ca-bundle.pem' - some environments require a custom CA to work correctly
I1104 11:27:09.018809       1 main.go:154] Starting the ConfigMap operator to monitor 'kube-cloud-config'
I1104 11:27:09.018853       1 controller.go:88] Starting node controller
I1104 11:27:09.018857       1 controller.go:91] Waiting for informer caches to sync for node workqueue
I1104 11:27:09.019276       1 controller.go:88] Starting configMap controller
I1104 11:27:09.019290       1 controller.go:91] Waiting for informer caches to sync for configMap workqueue
I1104 11:27:09.019310       1 controller.go:88] Starting cloud-private-ip-config controller
I1104 11:27:09.019314       1 controller.go:91] Waiting for informer caches to sync for cloud-private-ip-config workqueue
I1104 11:27:09.019479       1 controller.go:88] Starting secret controller
I1104 11:27:09.019487       1 controller.go:91] Waiting for informer caches to sync for secret workqueue
I1104 11:27:09.042435       1 controller.go:182] Assigning key: 6gks8w2-b5564-rw5lb-worker-0-wm9fx to node workqueue
I1104 11:27:09.042523       1 controller.go:182] Assigning key: 6gks8w2-b5564-rw5lb-worker-0-zfrj5 to node workqueue
I1104 11:27:09.042573       1 controller.go:182] Assigning key: 6gks8w2-b5564-rw5lb-master-0 to node workqueue
I1104 11:27:09.042638       1 controller.go:182] Assigning key: 6gks8w2-b5564-rw5lb-master-1 to node workqueue
I1104 11:27:09.042689       1 controller.go:182] Assigning key: 6gks8w2-b5564-rw5lb-master-2 to node workqueue
I1104 11:27:09.042735       1 controller.go:182] Assigning key: 6gks8w2-b5564-rw5lb-worker-0-flh8l to node workqueue
I1104 11:27:09.059024       1 controller.go:182] Assigning key: 10.0.129.10 to cloud-private-ip-config workqueue
I1104 11:27:09.119189       1 controller.go:96] Starting node workers
I1104 11:27:09.119297       1 controller.go:102] Started node workers
I1104 11:27:09.119365       1 controller.go:160] Dropping key '6gks8w2-b5564-rw5lb-worker-0-wm9fx' from the node workqueue
I1104 11:27:09.119423       1 controller.go:160] Dropping key '6gks8w2-b5564-rw5lb-worker-0-zfrj5' from the node workqueue
I1104 11:27:09.119469       1 controller.go:160] Dropping key '6gks8w2-b5564-rw5lb-master-0' from the node workqueue
I1104 11:27:09.119515       1 controller.go:160] Dropping key '6gks8w2-b5564-rw5lb-master-1' from the node workqueue
I1104 11:27:09.119580       1 controller.go:160] Dropping key '6gks8w2-b5564-rw5lb-master-2' from the node workqueue
I1104 11:27:09.120000       1 controller.go:96] Starting secret workers
I1104 11:27:09.120036       1 controller.go:102] Started secret workers
I1104 11:27:09.120056       1 controller.go:96] Starting cloud-private-ip-config workers
I1104 11:27:09.120078       1 controller.go:102] Started cloud-private-ip-config workers
I1104 11:27:09.130227       1 controller.go:160] Dropping key '10.0.129.10' from the cloud-private-ip-config workqueue
I1104 11:27:09.374836       1 node_controller.go:146] Setting annotation: 'cloud.network.openshift.io/egress-ipconfig: [{"interface":"508fc058-3044-459f-9e00-a42aab19b57b","ifaddr":{"ipv4":"10.0.0.0/16"},"capacity":{"ip":8}}]' on node: 6gks8w2-b5564-rw5lb-worker-0-flh8l
I1104 11:27:09.404368       1 controller.go:160] Dropping key '6gks8w2-b5564-rw5lb-worker-0-flh8l' from the node workqueue
[akaris@linux egressip]$ oc get nodes 6gks8w2-b5564-rw5lb-worker-0-flh8l -o yaml | grep egress
    cloud.network.openshift.io/egress-ipconfig: '[{"interface":"508fc058-3044-459f-9e00-a42aab19b57b","ifaddr":{"ipv4":"10.0.0.0/16"},"capacity":{"ip":8}}]'
    k8s.ovn.org/egress-assignable: ""
~~~

**Pull request merged**
